### PR TITLE
 Add premium sku related properties for managed environment

### PR DESCRIPTION
**Pull request by jijohn14**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [X] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [x] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following appy to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    - [ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [X] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/CommonDefinitions.json [hunk 1]
```diff
@@ -34,6 +34,19 @@
         }
       }
     },
+    "WorkloadProfileType": {
+      "description": "Workload profile type for container apps to execute.",
+      "enum": [
+        "General Purpose",
+        "Memory Optimized",
+        "Compute Optimized"
```
Comment (ruslany) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/CommonDefinitions.json:42]: >"Compute Optimized [](http://example.com/codeflow?start=8&length=18)

Is it ok to have spaces in these enums? Maybe it is - we just never had that before in this swagger (or in app service)

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 2]
```diff
@@ -660,6 +660,46 @@
...
     },
+    "EnvironmentSkuProperties": {
+      "description": "Managed Environment resource SKU properties.",
+      "type": "object",
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "Name of the Sku.",
+          "externalDocs": {
+            "url": "https://azure.microsoft.com/en-us/pricing/details/container-apps/"
```
Comment (ruslany) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:671]: >https://azure.microsoft.com/en-us/pricing/details/container-apps/ [](http://example.com/codeflow?start=20&length=65)

Should we maybe use aka.ms link there instead?

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 3]
```diff
@@ -841,6 +893,156 @@
...
+          "type": "integer",
+          "format": "int32",
+          "description": "The maximum capacity."
+        }
+      },
+      "required": [
+        "workloadProfileType"
+      ]
+    },
+    "Certificate": {
```
Comment (ruslany) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:925]: >Certificate [](http://example.com/codeflow?start=5&length=11)

Isn't this already defined in the commondefinitions?

Comment (jijohn14) [reply specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:925]: Fixed rebase issue.

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 4]
```diff
@@ -660,6 +660,46 @@
...
+        },
+        "capacity": {
+          "type": "integer",
+          "format": "int32",
+          "description": "Capacity of the SKU (number of deployed nodes of the SKU). For Consumption SKU capacity must be specified as 0."
+        }
+      },
+      "required": [
+        "name",
+        "capacity"
```
Comment (chinadragon0515) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:700]: If must be 0, could we make it optional?
And if user select premium, then create 2 app, one is MemoryOptimized and the other is ComputeOptiomized, then what we will do? set both node pool to this capacity?

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 5]
```diff
@@ -841,6 +893,34 @@
           "x-ms-secret": true
         }
       }
+    },
+    "WorkloadProfile": {
+      "description": "Description of a workload profile.",
```
Comment (chinadragon0515) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:898]: will this be an array ?

Comment (chinadragon0515) [reply specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:898]: And mismatch with file [ManagedEnvironments_CreateOrUpdate.json](https://github.com/Azure/azure-rest-api-specs/pull/19770/files#diff-6f12fcd602a58a7be78a0f0b9719c4b9aa35cb76966a2ca2dc12201129d941ba), line 22 which indicate an array via "["

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 6]
```diff
@@ -660,6 +660,46 @@
         }
       }
     },
+    "EnvironmentSkuProperties": {
+      "description": "Managed Environment resource SKU properties.",
```
Comment (chinadragon0515) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:664]: seems this definition is in middle, could we move down?

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 7]
```diff
@@ -796,6 +840,14 @@
                 "create",
                 "read"
               ]
+            },
+            "workloadProfiles": {
+              "description": "Workload profiles available for Managed Environment.",
+              "type": "array",
```
Comment (chinadragon0515) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:846]: is this readonly, if yes, how user set the max and min?

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/examples/ManagedEnvironments_CreateOrUpdate.json [hunk 8]
```diff
@@ -45,6 +90,10 @@
         "name": "testcontainerenv",
         "type": "Microsoft.App/managedEnvironments",
         "location": "East US",
+        "sku": {
+          "name": "Premium",
+          "capacity": 10
+        },
```
Comment (chinadragon0515) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/examples/ManagedEnvironments_CreateOrUpdate.json:96]: How 10 is related to below max and min number

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/examples/ManagedEnvironments_CreateOrUpdate.json [hunk 9]
```diff
@@ -35,7 +60,27 @@
               "customerId": "string"
             }
           },
-          "zoneRedundant": true
+          "zoneRedundant": true,
+          "workloadProfiles": [
+            {
+              "workloadProfileType": "GeneralPurpose",
+              "provisioningState": "Succeeded",
```
Comment (chinadragon0515) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/examples/ManagedEnvironments_CreateOrUpdate.json:67]: this is for [ManagedEnvironments_CreateOrUpdate.json](https://github.com/Azure/azure-rest-api-specs/pull/19770/files#diff-6f12fcd602a58a7be78a0f0b9719c4b9aa35cb76966a2ca2dc12201129d941ba), so should no provisioningState

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json (hunk 6 above)
Comment (jijohn14) [reply specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:664]: Perhaps we should make this alphabetical eventually

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/examples/ManagedEnvironments_CreateOrUpdate.json (hunk 8 above)
Comment (jijohn14) [reply specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/examples/ManagedEnvironments_CreateOrUpdate.json:96]: This is total nodes available for the environment i.e. sum of workload profile max can be less than equal to cluster maximum

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json (hunk 4 above)
Comment (jijohn14) [reply specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json]: Removed `capacity` property. This was inspired from [arm sku api contract](https://microsoft.sharepoint.com/:w:/r/teams/azureresourcemanagerteam/_layouts/15/Doc.aspx?sourcedoc=%7bD68FD846-9EC2-4D20-925D-133626DD6D1B%7d&file=arm-sku-api.docx&action=default&mobileredirect=true&share=IQFG2I_Wwp4gTZJdEzYm3W0bATGtnXbTsaqMOdZ4ErrL-Jo&cid=b5d55c0f-c7f2-4739-8940-2a9c3193d8fb). Will bring it back if needed

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 10]
```diff
@@ -796,11 +800,22 @@
...
                 "read"
               ]
+            },
+            "workloadProfiles": {
+              "description": "Workload profiles configured for the Managed Environment.",
+              "type": "array",
+              "items": {
+                "$ref": "#/definitions/WorkloadProfile"
+              },
+              "x-ms-identifiers": []
```
Comment (ruslany) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:810]: >  "x-ms-identifiers": [] [](http://example.com/codeflow?start=12&length=24)

This should not be empty

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 11]
```diff
@@ -841,6 +856,63 @@
...
+    },
+    "WorkloadProfile": {
+      "description": "Workload profile to scope container app execution.",
+      "type": "object",
+      "properties": {
+        "workloadProfileType": {
+          "$ref": "./CommonDefinitions.json#/definitions/WorkloadProfileType",
+          "description": "Workload profile type for the workloads to run on."
+        },
+        "minimum": {
```
Comment (ruslany) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:868]: >minimum [](http://example.com/codeflow?start=9&length=7)

should this be called minimumCapacity?

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 12]
```diff
@@ -841,6 +856,63 @@
           "x-ms-secret": true
         }
       }
+    },
+    "WorkloadProfile": {
+      "description": "Workload profile to scope container app execution.",
```
Comment (ruslany) [inline specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:861]: >. [](http://example.com/codeflow?start=71&length=1)

None of the other descriptions end with "."

**Review by ruslany — CHANGES_REQUESTED** [review]
:clock1:

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json (hunk 12 above)
Comment (jijohn14) [reply specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:861]: I think it is inconsistent. :)
Refer 
- line 23, 24, 42, 63, 64 in ManagedEnvironments.json
- line 439, 453, 458 in ContainerApps.json

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json (hunk 11 above)
Comment (jijohn14) [reply specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:868]: Renamed to minimumCount

File: specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json [hunk 13]
```diff
@@ -723,6 +723,10 @@
         }
       ],
       "properties": {
+        "sku": {
+          "$ref": "#/definitions/EnvironmentSkuProperties",
```
[the remark this answers is not in the document]
Comment (jijohn14) [reply specification/app/resource-manager/Microsoft.App/preview/2022-06-01-preview/ManagedEnvironments.json:774]: It is the same definition but we only want the `name` property within `sku`.

**Comment by qiaozha** [comment]
@jianyexi Can you take a look at this lintdiff error ? Thanks

**Comment by jianyexi** [comment]
> @jianyexi Can you take a look at this lintdiff error ? Thanks
Please ignore it, it's false alarm

**Review by ruslany — APPROVED** [review]
:shipit:

**Review by tadelesh — APPROVED** [review]

**Pull request merged**